Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -432,6 +439,73 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
     m_damage = Damage();
 }
 
+void TextureMapperLayer::collectDamageSelfChildrenReplicaFilterAndMask(TextureMapperPaintOptions& options, Damage& damage)
+{
+    bool hasFilterOrMask = [&] {
+        if (hasFilters())
+            return true;
+        if (m_state.maskLayer)
+            return true;
+        if (m_state.replicaLayer && m_state.replicaLayer->m_state.maskLayer)
+            return true;
+        return false;
+    }();
+    auto collectDamageSelfAndOthers = [&]() {
+        if (hasFilterOrMask)
+            collectDamageSelfChildrenFilterAndMask(options, damage);
+        else
+            collectDamageSelfAndChildren(options, damage);
+    };
+
+    if (m_state.replicaLayer) {
+        SetForScope scopedReplicaLayer(options.replicaLayer, this);
+        SetForScope scopedTransform(options.transform, options.transform);
+        options.transform.multiply(replicaTransform());
+        collectDamageSelfAndOthers();
+    }
+    collectDamageSelfAndOthers();
+}
+
+void TextureMapperLayer::collectDamageSelfChildrenFilterAndMask(TextureMapperPaintOptions& options, Damage& damage)
+{
+    Region overlapRegion;
+    Region nonOverlapRegion;
+    auto mode = ComputeOverlapRegionMode::Union;
+    if (m_state.maskLayer || (options.replicaLayer == this && m_state.replicaLayer->m_state.maskLayer))
+        mode = ComputeOverlapRegionMode::Mask;
+    ComputeOverlapRegionData data {
+        mode,
+        options.textureMapper.clipBounds(),
+        overlapRegion,
+        nonOverlapRegion
+    };
+    data.clipBounds.move(-options.offset);
+    computeOverlapRegions(data, options.transform, false);
+    ASSERT(nonOverlapRegion.isEmpty());
+
+    auto rects = overlapRegion.rects();
+    static const size_t OverlapRegionConsolidationThreshold = 4;
```

Comment:
Any good reason for the 4? I'd name it 'overlapRegion...'

Reply:
`constexpr size_t`?

Reply:
This code is a copy-paste from another place. I've extracted the common part now.